system-addon/content-src/components/TopSites/TopSites.jsx

```diff
@@ -331,14 +371,194 @@ class TopSitesEdit extends React.Component {
 
+class TopSiteForm extends React.Component {
+  constructor(props) {
+    super(props);
+    this.state = {
+      label: props.label || "",
+      url: props.url || "",
+      validationError: false
+    };
+    this.onLabelChange = this.onLabelChange.bind(this);
+    this.onUrlChange = this.onUrlChange.bind(this);
+    this.onCancelButtonClick = this.onCancelButtonClick.bind(this);
+    this.onAddButtonClick = this.onAddButtonClick.bind(this);
+    this.onSaveButtonClick = this.onSaveButtonClick.bind(this);
+    this.onUrlInputMount = this.onUrlInputMount.bind(this);
+  }
+  onLabelChange(event) {
+    this.resetValidation();
+    this.setState({"label": event.target.value});
+  }
+  onUrlChange(event) {
+    this.resetValidation();
+    this.setState({"url": event.target.value});
+  }
+  onCancelButtonClick(ev) {
+    ev.preventDefault();
+    this.props.onClose();
+  }
+  onAddButtonClick(ev) {
+    ev.preventDefault();
+    if (this.validateForm()) {
+      let site = {url: this.cleanUrl()};
+      if (this.state.label !== "") {
+        site.label = this.state.label;
+      }
+      this.props.dispatch(ac.SendToMain({
+        type: at.TOP_SITES_ADD,
+        data: {site}
+      }));
+      this.props.dispatch(ac.UserEvent({
+        source: TOP_SITES_SOURCE,
+        event: "TOP_SITES_ADD"
+      }));
+      this.props.onClose();
+    }
+  }
+  onSaveButtonClick(ev) {
+    ev.preventDefault();
+    if (this.validateForm()) {
+      let site = {url: this.cleanUrl()};
+      if (this.state.label !== "") {
+        site.label = this.state.label;
+      }
+      this.props.dispatch(ac.SendToMain({
+        type: at.TOP_SITES_EDIT,
+        data: {site, index: this.props.index}
+      }));
+      this.props.dispatch(ac.UserEvent({
+        source: TOP_SITES_SOURCE,
+        event: "TOP_SITES_EDIT",
+        action_position: this.props.index
+      }));
+      this.props.onClose();
+    }
+  }
+  cleanUrl() {
+    let url = this.state.url;
+    // If we are missing a protocol, prepend http://
+    if (!url.startsWith("http")) {
+      url = `http://${url}`;
+    }
+    return url;
+  }
+  resetValidation() {
+    if (this.state.validationError) {
+      this.setState({validationError: false});
+    }
+  }
+  validateUrl() {
+    const pattern = new RegExp("^(https?:\\/\\/)?" + // protocol
+      "((([a-z\\d]([a-z\\d-]*[a-z\\d])*)\\.)+[a-z]{2,}|" + // domain name
+      "((\\d{1,3}\\.){3}\\d{1,3}))" + // OR ip (v4) address
+      "(\\:\\d+)?(\\/[-a-z\\d%_.~+]*)*" + // port and path
+      "(\\?[;&a-z\\d%_.~+=-]*)?" + // query string
+      "(\\#[\\/-a-z\\d_]*)?$", "i"); // fragment locater
+    return pattern.test(this.state.url);
+  }
+  validateForm() {
+    this.resetValidation();
+    // Only the URL is required and must be valid.
+    if (!this.state.url || !this.validateUrl()) {
+      this.setState({validationError: "url"});
+      this.inputUrl.focus();
+      return false;
+    }
+    return true;
+  }
+  onUrlInputMount(input) {
```

Comment:
I don't see the refs being used in the tests, there's a couple other refs in the jsx as well

Reply:
it's just so I can get a reference to the `<input/>` element so that later I can `.focus()` it. In the past you could do `ref="foo"` then something like `this.refs.foo.focus()` but react took that away so now you have to do these gymnastics on mount to keep a reference to the underlying DOM nodes. Search has something similar which is how I learned to do it: https://github.com/mozilla/activity-stream/blob/master/system-addon/content-src/components/Search/Search.jsx#L24

Reply:
Ahh I see. Ok cool. Can you just remove the references to ```ref``` that are lying around in the code? Probably from the test pilot stuff

Reply:
yeah. we can kill all the `ref`s !! 🔪